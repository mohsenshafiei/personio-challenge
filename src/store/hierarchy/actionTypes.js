export default {
  INIT_APP: 'INIT_APP',
  INIT_APP_OK: 'INIT_APP_OK',
  JSON_FILE_UPLOADED: 'JSON_FILE_UPLOADED',
  JSON_FILE_UPLOADED_SUCCESSFUL: 'JSON_FILE_UPLOADED_SUCCESSFUL',
  REMOVE_PERSON: 'REMOVE_PERSON',
  ADD_PERSON: 'ADD_PERSON',
  CHANGE_FILTER: 'CHANGE_FILTER',
  TOGGLE_COLLAPSE: 'TOGGLE_COLLAPSE',
  DETECT_MULTIPLE_BOSS: 'DETECT_MULTIPLE_BOSS',
  REMOVE_MULTIPLE_PERSON: 'REMOVE_MULTIPLE_PERSON',
  NOTIFICATION: 'NOTIFICATION',
  CHANGE_HIERARCHY: 'CHANGE_HIERARCHY',
};
