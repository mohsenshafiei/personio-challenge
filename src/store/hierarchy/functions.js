export const updateEmployeesIds = (employees, parentId = '', parentLevel = 0) => employees.map((person, index) => {
  const id = `${parentId}_${index}`;
  const level = parentLevel + 1;
  return {
    ...person,
    id,
    level,
    employees: updateEmployeesIds(person.employees, id, level),
  };
});

export const filterEmployees = (employees, removedId) => employees.reduce((result, person) => {
  if (person.id === removedId) {
    return result;
  }
  if (person.employees && person.employees.length) {
    const filteredPerson = {
      ...person,
      employees: filterEmployees(person.employees, removedId),
    };
    return [...result, filteredPerson];
  }
  return [...result, person];
}, []);

export const addedEmployees = (employees, leaderId, newPerson) => employees.map((person) => {
  if (person.id === leaderId) {
    return {
      ...person,
      employees: [newPerson, ...person.employees],
    };
  }
  if (person.employees && person.employees.length) {
    return {
      ...person,
      employees: addedEmployees(person.employees, leaderId, newPerson),
    };
  }
  return person;
});

export const toggleCollapse = (employees, personId) => employees.map(person => ({
  ...person,
  collapsed: person.id === personId ? !person.collapsed : person.collapsed,
  employees: toggleCollapse(person.employees, personId),
}));

export const countFrequency = employees => employees.reduce((list, person) => {
  if (list[person]) {
    return {
      ...list,
      [person]: list[person] + 1,
    };
  }
  return {
    ...list,
    [person]: 1,
  };
}, {});

export const personsList = employees => employees.reduce((result, person) => {
  const count = personsList(person.employees);
  return [...result, person.name, ...count];
}, []);

export const transformEmployees = (employees, parentLevel = 0) => employees.map((person) => {
  const name = Object.keys(person)[0];
  const level = parentLevel + 1;
  return {
    name,
    level,
    ...person[name],
    employees: transformEmployees(person[name].employees || [], parentLevel),
  };
});

export const getPerson = (employees = [], personId) => {
  let employee = employees.find(person => person.id === personId);
  if (employee) {
    return employee;
  }
  for (let i = 0; i < employees.length; i += 1) {
    employee = getPerson(employees[i].employees, personId);
    if (employee) {
      return employee;
    }
  }
  return undefined;
};
