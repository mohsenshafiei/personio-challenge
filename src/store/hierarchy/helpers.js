export const updateEmployeesIds = (employees, parentId = '', parentLevel = 0) => employees.map((person, index) => {
  const id = parentId ? `${parentId}_${index}` : `${index}`;
  const level = parentLevel + 1;
  return {
    ...person,
    id,
    level,
    employees: updateEmployeesIds(person.employees, id, level),
  };
});

const getIdByPath = path => path.join('_');

const getPathById = id => id.split('_');

export const filterEmployees = (employees, removedId) => employees.reduce((result, person) => {
  if (person.id === removedId) {
    return result;
  }
  if (person.employees && person.employees.length) {
    const filteredPerson = {
      ...person,
      employees: filterEmployees(person.employees, removedId),
    };
    return [...result, filteredPerson];
  }
  return [...result, person];
}, []);

export const addedEmployees = (employees, leaderId, newPerson) => employees.map((person) => {
  if (person.id === leaderId) {
    return {
      ...person,
      employees: [newPerson, ...person.employees],
    };
  }
  if (person.employees && person.employees.length) {
    return {
      ...person,
      employees: addedEmployees(person.employees, leaderId, newPerson),
    };
  }
  return person;
});

export const toggleCollapse = (employees, personId) => {
  if (!personId) {
    return employees;
  }
  const personPath = getPathById(personId);
  const index = Number(personPath.shift());
  const person = employees[index];
  const isTargetPerson = personPath.length === 0;
  return [
    ...employees.slice(0, index),
    {
      ...person,
      collapsed: isTargetPerson ? !person.collapsed : person.collapsed,
      employees: isTargetPerson
        ? person.employees
        : toggleCollapse(person.employees, getIdByPath(personPath)),
    },
    ...employees.slice(index + 1),
  ];
};

export const countFrequency = employeesNames => employeesNames.reduce((list, personName) => {
  if (list[personName]) {
    return {
      ...list,
      [personName]: list[personName] + 1,
    };
  }
  return {
    ...list,
    [personName]: 1,
  };
}, {});

export const personsList = employees => employees.reduce((result, person) => {
  const count = personsList(person.employees);
  return [...result, person.name, ...count];
}, []);

export const transformEmployees = (employees, parentLevel = 0) => employees.map((person) => {
  const name = Object.keys(person)[0];
  const level = parentLevel + 1;
  return {
    name,
    level,
    ...person[name],
    employees: transformEmployees(person[name].employees || [], parentLevel),
  };
});

export const getPerson = (employees = [], personId) => {
  let employee = employees.find(person => person.id === personId);
  if (employee) {
    return employee;
  }
  for (let i = 0; i < employees.length; i += 1) {
    employee = getPerson(employees[i].employees, personId);
    if (employee) {
      return employee;
    }
  }
  return undefined;
};
