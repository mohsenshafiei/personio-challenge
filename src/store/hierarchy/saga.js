import {
  put, takeEvery, select,
} from 'redux-saga/effects';

const getPerson = (employees = [], personId) => {
  let employee = employees.find(person => person.id === personId);
  if (employee) {
    return employee;
  }
  for (let i = 0; i < employees.length; i += 1) {
    employee = getPerson(employees[i].employees, personId);
    if (employee) {
      return employee;
    }
  }
  return undefined;
};

function* hierarchyChanged(action) {
  if (action.leaderId.indexOf(action.personId) === 0) {
    yield put({ type: 'NOTIFICATION', title: 'This Action Is Not Possible, You Are Making a Loop!', style: 'error' });
  } else {
    const person = yield select(state => getPerson(state.hierarchy.employees, action.personId));
    if (person) {
      yield put({ type: 'REMOVE_PERSON', personId: action.personId });
      yield put({ type: 'ADD_PERSON', leaderId: action.leaderId, person });
    }
  }
}
function* removeMultiplePerson(action) {
  yield put({ type: 'REMOVE_PERSON', personId: action.personId });
  yield put({ type: 'DETECT_MULTIPLE_BOSS' });
}

function* fileUpload() {
  // you can send it to server here
  yield put({ type: 'JSON_FILE_UPLOADED_SUCCESSFUL', payload: true });
}

export default [
  takeEvery('JSON_FILE_UPLOADED', fileUpload),
  takeEvery('CHANGE_HIERARCHY', hierarchyChanged),
  takeEvery('REMOVE_MULTIPLE_PERSON', removeMultiplePerson),
];
