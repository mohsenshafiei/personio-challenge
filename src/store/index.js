/* eslint-disable no-undef,no-underscore-dangle */
import {
  createStore, applyMiddleware, combineReducers, compose,
} from 'redux';
import { createLogger } from 'redux-logger';
import createSagaMiddleware from 'redux-saga';
import hierarchyReducers from './hierarchy/reducers';
import mainReducers from './main/reducers';
import sagas from './sagas';

const reducers = combineReducers({
  main: mainReducers,
  hierarchy: hierarchyReducers,
});
const logger = createLogger();
const sagaMiddleware = createSagaMiddleware();
const composeEnhancers = typeof window === 'object' && window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__
  ? window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__({})
  : compose;

const middleware = applyMiddleware(logger, sagaMiddleware);
const enhancer = composeEnhancers(middleware);
const store = createStore(reducers, enhancer);

sagaMiddleware.run(sagas);
export default store;
