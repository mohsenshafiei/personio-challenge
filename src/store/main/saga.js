import { takeEvery } from 'redux-saga/effects';
import i18n from '../../i18n';

function changeLanguage() {
  if (i18n.language !== 'en') {
    i18n.lang = 'en';
    i18n.changeLanguage('en', (err) => {
      if (err) {
        return false;
      }
      window.localStorage.setItem('locale', 'en');
      return true;
    });
  } else {
    i18n.lang = 'de';
    i18n.changeLanguage('de', (err) => {
      if (err) {
        return false;
      }
      window.localStorage.setItem('locale', 'de');
      return true;
    });
  }
}

export default [
  takeEvery('CHANGE_LANGUAGE', changeLanguage),
];
