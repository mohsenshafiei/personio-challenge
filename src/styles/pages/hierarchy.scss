.hierarchy {
  position: absolute;
  top: 5em;
  right: 0;
  bottom: 0;
  left: 0;
  overflow-x: hidden;
  background-color: $personio-green;
  min-width: 20em;
  min-height: 30em;

  .list {
    width: 100%;
    position: absolute;
    margin-top: 1em;
  }
}
