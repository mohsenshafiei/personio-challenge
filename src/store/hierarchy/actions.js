import actionTypes from './actionTypes';

export const fileUpload = payload => (
  {
    type: actionTypes.JSON_FILE_UPLOADED,
    payload,
  }
);

export const changeHierarchy = (personId, leaderId) => (
  {
    type: actionTypes.CHANGE_HIERARCHY,
    personId,
    leaderId,
  }
);

export const changeFilter = filter => (
  {
    type: actionTypes.CHANGE_FILTER,
    filter,
  }
);

export const toggleCollapse = personId => (
  {
    type: actionTypes.TOGGLE_COLLAPSE,
    personId,
  }
);

export const detectMultipleBoss = () => (
  {
    type: actionTypes.DETECT_MULTIPLE_BOSS,
  }
);

export const removePerson = personId => (
  {
    type: actionTypes.REMOVE_MULTIPLE_PERSON,
    personId,
  }
);
