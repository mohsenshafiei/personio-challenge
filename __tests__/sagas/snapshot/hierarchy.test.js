import React from 'react';
import { expectSaga } from 'redux-saga-test-plan';
import { hierarchyChanged, removePerson } from '../../../src/store/hierarchy/saga';
import { store } from "../../../src/store/index";

describe('Change Hierarchy Snapshot Test', () => {
  const action1 = { type: 'CHANGE_HIERARCHY', personId: '00', leaderId: '001'};
  it('Hierarhcy change error snapshot', () => {
    return expectSaga(hierarchyChanged, action1)
      .run()
      .then((result) => {
        expect(result.toJSON()).toMatchSnapshot();
      });
  });

});

describe('Remove Person Snapshot Test', () => {
  let action = { type: 'REMOVE_MULTIPLE_PERSON', personId: '00'};
  it('remove multiple boss snapshot', () => {
    return expectSaga(removePerson, action)
      .run()
      .then((result) => {
        expect(result.toJSON()).toMatchSnapshot();
      });
  });
});
