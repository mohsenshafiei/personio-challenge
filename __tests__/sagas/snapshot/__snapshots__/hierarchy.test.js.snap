// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Change Hierarchy Snapshot Test Hierarhcy change error snapshot 1`] = `
Object {
  "put": Array [
    Object {
      "@@redux-saga/IO": true,
      "PUT": Object {
        "action": Object {
          "style": "error",
          "title": "This Action Is Not Possible, You Are Making a Loop!",
          "type": "NOTIFICATION",
        },
        "channel": null,
      },
    },
  ],
}
`;

exports[`Remove Person Snapshot Test remove multiple boss snapshot 1`] = `
Object {
  "put": Array [
    Object {
      "@@redux-saga/IO": true,
      "PUT": Object {
        "action": Object {
          "personId": "00",
          "type": "REMOVE_PERSON",
        },
        "channel": null,
      },
    },
    Object {
      "@@redux-saga/IO": true,
      "PUT": Object {
        "action": Object {
          "type": "DETECT_MULTIPLE_BOSS",
        },
        "channel": null,
      },
    },
  ],
}
`;
